tinygrad/engine/schedule.py

```diff
@@ -16,6 +16,73 @@
 
 BUF_LIMIT = {"METAL":32}
 
+# **** big graph spec
+
+tensor_uop_spec = PatternMatcher([
+  # ** stable and well understood specs
+  # DEVICE and BUFFER
+  (UPat(Ops.DEVICE, dtypes.void, (), name="device"), lambda device: isinstance(device.arg, str)),
+  (UPat(Ops.BUFFER, src=(UPat(Ops.DEVICE),), name="buf"), lambda buf:
+   # arg: (number, size)
+   isinstance(buf.arg, tuple) and len(buf.arg) == 2 and all(isinstance(x, int) for x in buf.arg) and \
+   # dtype
+   isinstance(buf.dtype, (DType, ImageDType))),
+  # movement ops
+  (UPat(GroupOp.Movement, name="mv", src=(UPat.var("x"),)), lambda mv,x: isinstance(mv.arg, tuple) and mv.dtype == x.dtype),
+  # tensor variable bindings
+  (UPat(Ops.BIND, dtype=dtypes.int, src=(UPat(Ops.DEFINE_VAR), UPat.cvar(dtype=dtypes.int)), arg=None), lambda: True),
+  # DETACH and CONTIGUOUS change how we interpret the source UOp
+  # CONTIGUOUS ensures the source UOp realizes to a device Buffer
+  (UPat((Ops.DETACH, Ops.CONTIGUOUS), name="root", src=(UPat.var("x"),), arg=None), lambda root,x: root.dtype == x.dtype),
+
+  # ** specs with room for refactoring and improving
+
+  # COPY
+  (UPat(Ops.COPY, name="copy", src=(UPat.var("copyin"),)), lambda copy,copyin:
+   # arg (device, clone?)
+   # TODO: use Ops.DEVICE src
+   isinstance(copy.arg, tuple) and len(copy.arg) == 2 and isinstance(copy.arg[0], str) and isinstance(copy.arg[1], bool) and \
+   # dtype
+   copy.dtype == copyin.dtype),
+
+  # VIEW(BUFFER) applies a ShapeTracker on top of the underlying device buffer
+  # NOTE: VIEW size exactly matches the underlying BUFFER, tensor doesn't apply movement ops to the VIEW
+  (UPat(Ops.VIEW, name="view", src=(UPat(Ops.BUFFER, name="buf"),)),
+   lambda view,buf: view.dtype == buf.dtype and view.size == buf.size and view.st.contiguous),
```

Comment:
Why does this VIEW exist?

Is it purely there to represent the shape of the on-device Buffer (aka. preserve STORE st), because of `tensor.shape`?


I added some asserts to restrict the ShapeTracker of this VIEW on BUFFER, maybe TIP 4 is a good place to define this
https://github.com/tinygrad/tinygrad/issues/8294.

movement ops on BUFFER look like EXPAND, RESHAPE, etc at the tensor level.
The scheduler is free to add up these sts, but the resulting VIEW(BUFFER) has a different meaning, `LOAD`.